vespalib/src/vespa/vespalib/net/tls/certificate_verification_callback.h

```diff
@@ -0,0 +1,28 @@
+// Copyright 2018 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
+#pragma once
+
+#include "peer_credentials.h"
+
+namespace vespalib::net::tls {
+
+// Verification callback invoked when a signed X509 certificate is presented
+// from a peer during TLS handshaking.
+// Only invoked for the leaf peer certificate, _not_ for any CAs (root or intermediate).
+// Only invoked iff the certificate has already passed OpenSSL pre-verification.
+struct CertificateVerificationCallback {
+    virtual ~CertificateVerificationCallback() = default;
+    // Return true iff the peer credentials pass verification, false otherwise.
+    // Must be thread safe.
+    virtual bool verify(const PeerCredentials& peer_creds) const = 0;
```

Comment:
Consider adding the remote IP address (or similar) as well. That will allow for the typical browser verification (CN/SAN matches DNS of remote ip).

Reply:
Definitely agree we want this, but the context in which the callback is invoked unfortunately does not contain any networking details. Will likely want the networking layer to explicitly pass down peer information when crypto sockets/codecs are created. Then this information can in turn be passed on to the verification callback.